### ISSUE TITLE
Add Diamond Price Prediction Web App

### ISSUE THREAD

**Issue by keshripritesh** [PRIYANSHU2026/DataSentience-AIML#774]
## 💎 [GSSoC’25] Add Diamond Price Prediction Web App

---

### 🧠 **Description**

This project introduces a **Diamond Price Prediction Web App** built using **Streamlit** and **Machine Learning**.
The app predicts diamond prices based on user-input features such as carat weight, cut, color, clarity, depth, table, and dimensions. It also provides insights into historical diamond data through a dataset view.

Users can interactively input diamond attributes and receive **predicted price estimates instantly**, making it a practical tool for both buyers and sellers.

---

### 🎯 **Objective**

To create an interactive web application that predicts diamond prices using a trained machine learning model while also allowing users to explore the underlying dataset.

---

### 🚀 **Key Features**

* 💎 **Diamond Price Prediction:** Predicts prices using user-provided diamond characteristics via a **Random Forest model**.
* 📊 **Dataset View:** Explore historical diamond data (`diamonds.csv`) in a tabular format.
* 📈 **Visualization & EDA:** Includes pie charts, violin plots, box plots, count plots, heatmaps, and model comparison graphs.
* 🧭 **Interactive Streamlit Interface:** User-friendly GUI for easy input and real-time predictions.

---

### 🧩 **Methodology**

1. **Data Preprocessing & Feature Engineering**
2. **Exploratory Data Analysis (EDA):** Visualization using Matplotlib, Seaborn, Pandas, and Plotly.
3. **Model Training & Evaluation:**

   * Random Forest: 98% accuracy
   * Support Vector Machine: 87% accuracy
   * Decision Tree: 97% accuracy
   * Linear Regression: 90% accuracy
   * 10-fold cross-validation on Random Forest: 98.25% average accuracy
4. **Deployment:** Model deployed via Streamlit to predict diamond prices in real-time.

---

### 🧩 **Tech Stack / Libraries**

* **Python**
* **Joblib** – Load trained Random Forest model
* **Scikit-learn** – Machine learning operations
* **Streamlit** – Web application GUI
* **Pandas / NumPy** – Data handling and numerical operations
* **Matplotlib / Seaborn / Plotly** – Data visualization

---

### ⚙️ **How to Run**

1. **Clone the Repository**:

   ```sh
   git clone <repository_url>
   ```
2. **Install Dependencies**:

   ```sh
   pip install -r requirements.txt
   ```
3. **Download the Pre-trained Model**:
   Drive link: [[Random Forest Model](https://drive.google.com/file/d/1OQ3WwkFdhlTmOptbfgudzCx6nS5QggQH/view?usp=sharing)](https://drive.google.com/file/d/1OQ3WwkFdhlTmOptbfgudzCx6nS5QggQH/view?usp=sharing)
   *(Keep it in the same directory as `app.py`)*
4. **Run the Application**:

   ```sh
   streamlit run app.py
   ```
5. **Interact with the App**: Predict diamond prices and explore the dataset.

---

### 🩹 **Troubleshooting**

* Ensure `diamonds.csv` and the pre-trained model are in the correct directory.
* Verify all dependencies are installed using `pip install -r requirements.txt`.
* Streamlit should be updated to the latest version for compatibility.

---

### 📜 **License**

Licensed under the **MIT License**.

---
